src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/PhysicalAddress.cs

```diff
@@ -105,70 +105,90 @@ public byte[] GetAddressBytes()
 
         public static PhysicalAddress Parse(string address)
         {
-            int validCount = 0;
-            bool hasDashes = false;
-            byte[] buffer = null;
+            int validSegmentLength;
+            char? delimiter = null;
+            byte[] buffer;
 
             if (address == null)
             {
-                return PhysicalAddress.None;
+                return None;
             }
 
-            // Has dashes?
             if (address.Contains('-'))
             {
-                hasDashes = true;
-                buffer = new byte[(address.Length + 1) / 3];
-
                 if ((address.Length + 1) % 3 != 0)
                 {
-                    throw new FormatException(SR.Format(SR.net_bad_mac_address, address));
+                    ThrowBadAddressException(address);
+                }
+
+                delimiter = '-';
+                buffer = new byte[(address.Length + 1) / 3]; // allow any length that's a multiple of 3
+                validSegmentLength = 2;
+            }
+            else if (address.Contains(':'))
+            {
+                delimiter = ':';
+                validSegmentLength = GetValidSegmentLength(address, ':');
+                if (validSegmentLength != 2 && validSegmentLength != 4)
+                {
+                    ThrowBadAddressException(address);
+                }
+                buffer = new byte[6];
+            }
+            else if (address.Contains('.'))
+            {
+                delimiter = '.';
+                validSegmentLength = GetValidSegmentLength(address, '.');
+                if (validSegmentLength != 4)
+                {
+                    ThrowBadAddressException(address);
                 }
+                buffer = new byte[6];
             }
             else
             {
                 if (address.Length % 2 > 0)
                 {
-                    throw new FormatException(SR.Format(SR.net_bad_mac_address, address));
+                    ThrowBadAddressException(address);
                 }
 
+                validSegmentLength = address.Length;
                 buffer = new byte[address.Length / 2];
             }
 
+            int validCount = 0;
             int j = 0;
             for (int i = 0; i < address.Length; i++)
             {
-                int value = (int)address[i];
+                int value = address[i];
 
-                if (value >= 0x30 && value <= 0x39)
+                if (value >= '0' && value <= '9')
+                {
+                    value -= '0';
+                }
+                else if (value >= 'A' && value <= 'F')
                 {
-                    value -= 0x30;
+                    value -= ('A' - 10);
                 }
-                else if (value >= 0x41 && value <= 0x46)
+                else if (value >= 'a' && value <= 'f')
```

Comment:
I wonder why it is not the pattern:
```c#
                else if ((value - 'a') <= ('f' - 'a'))
```

Reply:
or `value <= 'f'`. JIT ASM: [comparison](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAZgAJiAmcgYQFgAoAbyfPcouAggBtyAsgEYAFGAAW2KOQBu2XgFcYASnIBeAHyz5S8gB415AOQAzIwG42HMuW59BVMZOlzFK9VpGvdcY9iOqBuQipkbkvkb+ypaMHJy2PPwCpE5S2m6qmum6GoaRYQBkBdkw+nlmMQC+QA==).

Reply:
actually it will not do the proper 'a'..'f' range check. e.g. `value = 'A'` it fails.